### PR TITLE
Added super init to Module

### PR DESCRIPTION
**Pull request by Vivswan**
Added super init to Module for complex user modules derived from multiple python classes.
And by adding the super __init__ call at the end so it doesn't change any functionality of Module class. 

I am working on building a module for simulating analog neural network on PyTorch.
and this small change is really useful for that and we can definitely think of many other useful cases especially for more module or mro hierarchy.

Issues: https://github.com/pytorch/pytorch/issues/28746, https://github.com/pytorch/pytorch/issues/48626, https://github.com/pytorch/pytorch/issues/61662, https://github.com/pytorch/pytorch/issues/74036

### CODE REVIEW

**Review by albanD — COMMENTED** [review]
We tried this before and it was reverted: https://github.com/pytorch/pytorch/issues/74036
I can check what is the status now.

**Comment by Vivswan** [comment]
I see (https://github.com/pytorch/pytorch/issues/74036). 

I under your opinion on providing args and kwargs to the super class. 
Till we figure out the proper functionality to do so. 
We can probably still add `super().__init__()` without arguments which will provide usefulness to majority of the developers dealing with this issue.

File: torch/nn/modules/module.py [hunk 1]
```diff
@@ -461,6 +461,7 @@ def __init__(self) -> None:
         super().__setattr__('_load_state_dict_pre_hooks', OrderedDict())
         super().__setattr__('_load_state_dict_post_hooks', OrderedDict())
         super().__setattr__('_modules', OrderedDict())
+        super().__init__()
```
Comment (albanD) [inline torch/nn/modules/module.py:464]: This still doesn't allow you to do things like:
```python
import torch
from torch import nn

class MyClass:
    def __init__(self, foo):
        self.foo = foo


class MyMod(nn.Module, MyClass):
    def __init__(self, foo):
        super().__init__(foo)

m = MyMod(3)
print(m)
```

Could you change the signature to be `    def __init__(self, *args, **kwargs) -> None:` and this call to `super().__init__(*args, **kwargs)` please?

**Review by albanD — COMMENTED** [review]
In general, this is BC-breaking for anyone that is currently doing mixin of nn.Module and working around the lack of super() init() manually...
I guess that is ok to break as it is an explicit error

File: torch/nn/modules/module.py (hunk 1 above)
Comment (Vivswan) [reply torch/nn/modules/module.py]: done in https://github.com/pytorch/pytorch/pull/91819/commits/e60b22dede072ef04999ca9bd199898976f161ed

**Comment by Vivswan** [comment]
if approved, then it can be in the docs of `nn.Module` as an important notice that now it does call `super.__init__(*,**)`

OR 

to help everyone with huge library's and any error due to this change.
we can provide an option with which they can stop the call to `super().__init__(*,**)`.
this option can be deprecated in the future. 

Something maybe like this:

```python
class Module:
    _call_super: bool = True #deprecated in future
    
    def __init__(self, *args, **kwargs) -> None:
        ......
        ......
        if Module._call_super: 
            super().__init__(*args, **kwargs)
```

in this way anyone can disable the `super().__init__(*,**)` just by
```python
import torch
torch.nn.Module._call_super = False
```

**Comment by albanD** [comment]
I like this idea of having a class attribute to control this behavior as a way to make the transition smoother.

Could you add this new class attribute and make it False by default.
That will make this PR not break backward compatibility but allow you to enable this if you need this.

We can do a follow up, BC-breaking, PR that switch the default behavior to True.

**Comment by Vivswan** [comment]
Added `call_super_init: bool = False` to `nn.Module`

File: torch/nn/modules/module.py [hunk 2]
```diff
@@ -462,6 +463,9 @@ def __init__(self) -> None:
         super().__setattr__('_load_state_dict_post_hooks', OrderedDict())
         super().__setattr__('_modules', OrderedDict())
 
+        if Module.call_super_init:
```
Comment (albanD) [inline torch/nn/modules/module.py:466]: You should use `self.call_super_init` here such that subclasses can override the behavior just for them without changing the global Module behavior!

Comment (Vivswan) [reply torch/nn/modules/module.py]: ok,
https://github.com/pytorch/pytorch/pull/91819/commits/03b24eb3c90bf51af38eade867f2d1d8c531200e

**Review by albanD — COMMENTED** [review]
The change looks good to me. Could you just add a test in `test/test_nn.py` to make sure it does the right thing when a subclass sets this flag to True please and then we can merge this!

**Comment by Vivswan** [comment]
Can you give me a link to documentation, how to create tests for pytorch?

**Comment by albanD** [comment]
Ho I'm not sure we have doc for that :/
You can add a new function here: https://github.com/pytorch/pytorch/blob/cf5a40c2b4edbf2c17ce4e6bd52704d15c65aead/test/test_nn.py#L165

Call it `test_module_super_init` and check that you do get the behavior you want.

**Comment by albanD** [comment]
The test failures are relevant. Not sure why at a glance...

**Comment by Vivswan** [comment]
All the failed test are coming from the following code, since now we do accept `*args, **kwargs` in `nn.Module`.

https://github.com/pytorch/pytorch/blob/6f1727b2884808399d6e50444a4f4c04284553f1/test/test_module_init.py#L416-L419

Either we can change this test or we can throw a `TypeError` if args or kwargs are not empty and call_super_init is false

**Comment by albanD** [comment]
Ho that is interesting haha
Raising an error when call_super_init is `False` sounds good.
If you could raise the same error (as much as possible, including type and messagee) that is raised today that would be best so that we don't have to update the tests (and thus reduce the chance of breaking user code).

**Comment by Vivswan** [comment]
Do you know why remaining test are failing?, i don't think it has to with any of of the changes in this pr, we can rebase or merge master into this pr, but master branch is also failing some test right now.

**Review by albanD — APPROVED** [review]

**Pull request closed without merging**